### PR TITLE
Basic Adiabatic Valve Model

### PR DESCRIPTION
**Pull request by eslickj**
This adds a basic adiabatic valve model.  I'm trying to move the generic valve model out of the power plants library.  The pressure-flow relation by default, assumes the flow coefficient is for a specific liquid and there is no specific gravity explicitly in the pressure-flow equation.  If you look at the most typical valve model this is just combining the specific gravity with the flow coefficient.  My goal is that once the basic model is approved and good we can add a larger library of valve types by providing a selection of pressure-flow equations for types of valves and types of fluid over time.

I suspect the docs are riddled with typos that I'm blind to at this point, so I also give it a couple more reviews.  Other than that should be all there. 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/core/util/testing.py [hunk 1]
```diff
@@ -79,30 +80,31 @@ def initialization_tester(m, dof=0, **init_kwargs):
     Raises:
         AssertionErrors is an issue is found
```
Comment (andrewlee94) [inline idaes/core/util/testing.py:81]: Not your fault, but this should be "if an issue is found"

File: idaes/core/util/testing.py [hunk 2]
```diff
@@ -79,30 +80,31 @@ def initialization_tester(m, dof=0, **init_kwargs):
...
     """
+    unit = getattr(m.fs, unit_name)
```
Comment (andrewlee94) [inline idaes/core/util/testing.py:83]: Whilst this allows a custom unit name, it still assumes `m.fs.` i the flowsheet object. If we are going to allow this flexibility, then maybe we should go all the way and make "unit" a required input with the actual model object.

File: idaes/generic_models/unit_models/valve.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""
+This provides standard valve models for adaibatice control valves.  Beyond the
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/valve.py:14]: Typo: adiabatic

File: idaes/generic_models/unit_models/valve.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    def pressure_flow_equation(b2, t):
+        Po = b2.control_volume.properties_out[t].pressure
+        Pi = b2.control_volume.properties_in[t].pressure
+        F = b2.control_volume.properties_in[t].flow_mol
+        Cv = b2.Cv
+        fun = b2.valve_function[t]
+        return F ** 2 == Cv ** 2 * (Pi - Po) * fun ** 2
+
+
+def _define_config(config):
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/valve.py:103]: Not that it matters, but it is a little non-standard to have this as a separate function (i.e. none of the other core unit models do this).

File: idaes/generic_models/unit_models/valve.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        doc="Valve flow coefficent",
+        units=umeta("amount")/umeta("time")/umeta("pressure")**0.5
+    )
+    b.Cv.fix()
+
+    b.flow_var = pyo.Reference(b.control_volume.properties_in[:].flow_mol)
+    b.pressure_flow_equation_scale = lambda x : x**2
+
+    @b.Constraint(b.flowsheet().config.time)
+    def pressure_flow_equation(b2, t):
```
Comment (andrewlee94) [inline idaes/generic_models/unit_models/valve.py:94]: Whilst I know why you did it, the use of `b2` looks a little weird.

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
I saw a couple of typos, but otherwise looks good.

File: idaes/core/util/testing.py (hunk 2 above)
Comment (eslickj) [reply idaes/core/util/testing.py:83]: Sure.  That's no problem.  Will change.

File: idaes/generic_models/unit_models/valve.py (hunk 4 above)
Comment (eslickj) [reply idaes/generic_models/unit_models/valve.py:103]: Ah, yeah.  I'll move it to the normal place.

File: idaes/generic_models/unit_models/valve.py (hunk 5 above)
Comment (eslickj) [reply idaes/generic_models/unit_models/valve.py:94]: I thought so too, but I couldn't think of anything less weird.  Maybe I'll change the top level b to "valve" or something more descriptive.

File: docs/technical_specs/model_libraries/generic/unit_models/valve.rst [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+Custom Valve Functions
+~~~~~~~~~~~~~~~~~~~~~~
+
+In general, the valve opening should be restricted to range from 0 to 1.  The valve
+function should be a named expression attached to the valve model called
+``valve_function`` which takes the valve opening and computes a value that goes
+from approximately zero when valve opening is 0 to 1 when the valve opening is one.
+The valve function can have parameters as needed, so custom valve functions are
+defined in a callback so any additional parameters can be added in the callback.
```
Comment (blnicho) [inline docs/technical_specs/model_libraries/generic/unit_models/valve.rst:112]: I suggest rephrasing this sentence to: "Custom valve functions are defined using a callback function." 

The point about being able to include additional parameters is made in the next paragraph.

File: docs/technical_specs/model_libraries/generic/unit_models/valve.rst [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+  F^2 = C_v^2\left(P_{in} - P_{out}\right)f(x)^2
+
+Custom Pressure Flow Relations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Other pressure-flow equations can be specified via callback supplied to the unit
+configuration option ``pressure_flow_callback``. The callback allows both the form
+and flow basis of the pressure-flow equation to be specified.
+
+The callback can added parameters and variables as needed. It is recommended that
```
Comment (blnicho) [inline docs/technical_specs/model_libraries/generic/unit_models/valve.rst:156]: Typo: "The callback can **add** ..."

File: docs/technical_specs/model_libraries/generic/unit_models/valve.rst [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+          F = b2.control_volume.properties_in[t].flow_mol
+          Cv = b2.Cv
+          fun = b2.valve_function[t]
+          return F ** 2 == Cv ** 2 * (Pi - Po) * fun ** 2
+
+
+Initialization
+--------------
+
+This just calls the initialization routine from PressureChanger. Either and
```
Comment (blnicho) [inline docs/technical_specs/model_libraries/generic/unit_models/valve.rst:209]: Typo: "Either **an** ..."

File: idaes/generic_models/unit_models/valve.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    config.declare(
+        "valve_function_callback",
+        ConfigValue(
+            default=ValveFunctionType.linear,
+            description="Valve function type or callback for custom",
+            doc="""This takes either an enumerated valve function type in: {
+ValveFunctionType.linear, ValveFunctionType.quick_opening,
+ValveFunctionType.equal_percentage, ValveFunctionType.custom} or a callback
+function that takes a valve model object as an argument and adds a time-indexed
+valve_function expression to it. Any addtional required variables, expressions,
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:125]: Typo: additional

File: idaes/generic_models/unit_models/valve.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+or constraints required can also be added by the callback.""",
+        ),
+    )
+    config.declare(
+        "pressure_flow_callback",
+        ConfigValue(
+            default=pressure_flow_default_callback,
+            description="Callback function providing the valve_function expression",
+            doc="""This callback function takes a valve model object as an argument
+an adds a time-indexed valve_function expression to it.  Any addtional required
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:135]: Typo: "**and** adds a ..."
Typo: additional

File: idaes/generic_models/unit_models/valve.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        super().initialize(
+            state_args=state_args, outlvl=outlvl, solver=solver, optarg=optarg
+        )
+
+        # reload original spec
+        from_json(self, sd=istate, wts=sp)
+
+    def calculate_scaling_factors(self):
+        """
+        Calculte pressure flow constraint scaling from flow variable scale.
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:242]: Typo: calculate

File: idaes/generic_models/unit_models/valve.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+        """
+        # The valve of the valve opening and the output of the valve function
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:244]: Typo: "The **value** of the valve opening..."

File: idaes/generic_models/unit_models/valve.py [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+        # expression are between 0 and 1, so the only thing that needs to be
+        # scaled here is the pressure-flow constraint, which can be scaled by
+        # using the flow varable scale.  The flow variable could be defined
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:247]: Typo: variable

File: idaes/generic_models/unit_models/valve.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+        # in differnt ways, so the flow variable is determined here from a
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:248]: Typo: different

File: idaes/generic_models/unit_models/valve.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+        # "flow_var[t]" reference set in the pressure-flow callback. The flow
+        # term could be in verious forms, so an optional
```
Comment (blnicho) [inline idaes/generic_models/unit_models/valve.py:250]: Typo: various

**Review by blnicho — COMMENTED** [review]
I found a few typos

**Review by andrewlee94 — DISMISSED** [review]
Looks OK to me.

**Comment by eslickj** [comment]
Okay, I think it's all ready to go again.  Typos fixed. Looks like the test coverage went down a good bit, but I think the valve test coverage is complete.  Not sure what to make of it.

**Comment by lbianchi-lbl** [comment]
> Okay, I think it's all ready to go again. Typos fixed. Looks like the test coverage went down a good bit, but I think the valve test coverage is complete. Not sure what to make of it.

@eslickj The apparent large drop in coverage reported by Codecov happens when the integration tests are not run for a particular commit. Once the integration tests are run, the Codecov report in the PR comment is updated with the newest values. This behavior is confusing, and I'm working to change it; in the meanwhile, the integration tests can be triggered manually when necessary by adding the "approved" label (which might require removing it first if already present). Sorry again for any confusion due to this.

**Review by blnicho — APPROVED** [review]

**Comment by eslickj** [comment]
@lbianchi-lbl, there are a couple files like ver.py where the coverage went down, but I didn't do anything related to them.  Even though my patch have better coverage than the overall project, somehow the coverage still went down.

**Comment by lbianchi-lbl** [comment]
@eslickj I looked into this more carefully, and I see what you mean about the coverage drop in existing files such as `ver.py`: this is independent from the partial coverage update due to the integration tests, and seems to be due to an actual difference in the tests between the two runs being compared.

Judging from [this](https://codecov.io/gh/IDAES/idaes-pse/compare/f3c2a108f7f8ce8a555f342bf97ed103d366bb7d...19633f6d58fff76b38bcb3040b01a15e064aa93c/src/idaes/ver.py) part of the Codecov report, my guess is that in the test run used for the base (i.e. the one that was run on `main`), the code path where `git_head` was found was taken, while in the test run for the HEAD of the PR branch, the code path where `git_head` kept its fallback value of `None` was taken instead. This in turn could be due to a non-deterministic cause, e.g. in the hypothesis that the git parsing succeeds on Linux and fails on Windows, the latest test to finish would overwrite the previous result.

I am not familiar how Codecov deals with this type of conflicts for multiple runs covering the same code, and so this is just a guess at this point. If this is the case, I would consider this to be rather on the side of "false positive" in terms of the coverage check, unless we'd like to pinpoint the causes of the discrepancy more accurately.

(I'm attaching a screenshot of the coverage diff I'm referring to since the link doesn't seem to be working reliably)
image

**Review by andrewlee94 — APPROVED** [review]
Looks good.

**Pull request merged**